iothub/device/src/Transport/AmqpIoT/AmqpIoTSendingLink.cs

```diff
@@ -0,0 +1,153 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.Azure.Amqp;
+using Microsoft.Azure.Amqp.Framing;
+using Microsoft.Azure.Devices.Shared;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.Devices.Client.Transport.AmqpIoT
+{
+    internal class AmqpIoTSendingLink
+    {
+        public event EventHandler Closed;
+        private SendingAmqpLink _sendingAmqpLink;
+
+        public AmqpIoTSendingLink(SendingAmqpLink sendingAmqpLink)
+        {
+            _sendingAmqpLink = sendingAmqpLink;
+            _sendingAmqpLink.Closed += _sendingAmqpLinkClosed;
+        }
+
+        private void _sendingAmqpLinkClosed(object sender, EventArgs e)
+        {
+            if (Logging.IsEnabled) Logging.Enter(this, $"{nameof(_sendingAmqpLinkClosed)}");
+            Closed.Invoke(sender, e);
+        }
+
+        internal Task CloseAsync(TimeSpan timeout)
+        {
+            if (Logging.IsEnabled) Logging.Enter(this, $"{nameof(CloseAsync)}");
+            return _sendingAmqpLink.CloseAsync(timeout);
+        }
+
+        internal void Abort()
+        {
+            if (Logging.IsEnabled) Logging.Enter(this, $"{nameof(Abort)}");
+            _sendingAmqpLink.Abort();
+        }
+
+        #region Telemetry handling
+        internal async Task<AmqpIoTOutcome> SendMessageAsync(Message message, TimeSpan timeout)
+        {
+            if (Logging.IsEnabled) Logging.Enter(this, message, $"{nameof(SendMessageAsync)}");
+
+            AmqpMessage amqpMessage = AmqpIoTMessageConverter.MessageToAmqpMessage(message);
+            Outcome outcome = await SendAmqpMessageAsync(amqpMessage, timeout).ConfigureAwait(false);
+            return new AmqpIoTOutcome(outcome);
+        }
+
+        internal async Task<AmqpIoTOutcome> SendMessagesAsync(IEnumerable<Message> messages, TimeSpan timeout)
+        {
+            if (Logging.IsEnabled) Logging.Enter(this, $"{nameof(SendMessagesAsync)}");
+
+            // List to hold messages in Amqp friendly format
+            var messageList = new List<Data>();
+
+            foreach (Message message in messages)
+            {
+                using (AmqpMessage amqpMessage = AmqpIoTMessageConverter.MessageToAmqpMessage(message))
+                {
+                    var data = new Data()
+                    {
+                        Value = amqpMessage.DataBody
+                    };
+                    messageList.Add(data);
+            }
+            }
+
+            Outcome outcome;
+            using (AmqpMessage amqpMessage = AmqpMessage.Create(messageList))
+            {
+                amqpMessage.MessageFormat = AmqpConstants.AmqpBatchedMessageFormat;
+                outcome = await SendAmqpMessageAsync(amqpMessage, timeout).ConfigureAwait(false);
+            }
+
+            AmqpIoTOutcome amqpIoTOutcome = new AmqpIoTOutcome(outcome);
+            amqpIoTOutcome.ThrowIfNotAccepted();
```

Comment:
This can result in nullref exception instead (not good), based on the code below (extracted from somewhere in this PR):

```
        private Outcome Outcome; 

        public AmqpIoTOutcome(Outcome outcome) 
        { 
            Outcome = outcome; 
        } 

        public void ThrowIfNotAccepted() 
        { 
            if (Outcome.DescriptorCode != Accepted.Code) 
            { 
                throw AmqpIoTErrorAdapter.GetExceptionFromOutcome(Outcome); 
            } 

        }
```

Reply:
Fixed

---
In reply to: [293026125](https://github.com/Azure/azure-iot-sdk-csharp/pull/947#discussion_r293026125) [](ancestors = 293026125)